lib/SILOptimizer/Analysis/ClosureScope.cpp

```diff
@@ -0,0 +1,196 @@
+  // closure. The indices index into the `indexedScopes` vector. If the indexed
+  // scope is nullptr, then that function has been deleted.
+  using ClosureScopes = llvm::SmallVector<int, 1>;
+
+  // Map each closure to its parent scopes.
+  llvm::DenseMap<SILFunction *, ClosureScopes> closureToScopesMap;
+
+public:
+  void reset() {
+    indexedScopes.clear();
+    scopeToIndexMap.clear();
+    closureToScopesMap.clear();
+  }
+
+  void erase(SILFunction *F) {
+    // If this function is a mapped closure scope, remove it, leaving a nullptr
+    // sentinel.
+    auto indexPos = scopeToIndexMap.find(F);
+    if (indexPos != scopeToIndexMap.end()) {
+      indexedScopes[indexPos->second] = nullptr;
+      scopeToIndexMap.erase(F);
+    }
+    // If this function is a closure, remove it.
+    closureToScopesMap.erase(F);
+  }
+
+  // Record all closure scopes in this module.
+  void compute(SILModule *M);
+
+  bool isClosureScope(SILFunction *F) { return scopeToIndexMap.count(F); }
+
+  // Return a range of scopes for the given closure. The elements of the
+  // returned range have type `SILFunction *` and are non-null. Return an empty
+  // range for a SILFunction that is not a closure or is a dead closure.
+  ScopeRange getClosureScopes(SILFunction *ClosureF) {
+    IndexRange indexRange(nullptr, nullptr);
+    auto closureScopesPos = closureToScopesMap.find(ClosureF);
+    if (closureScopesPos != closureToScopesMap.end()) {
+      auto &indexedScopes = closureScopesPos->second;
+      indexRange = IndexRange(indexedScopes.begin(), indexedScopes.end());
+    }
+    return makeOptionalTransformRange(indexRange,
+                                      IndexLookupFunc(indexedScopes));
+  }
+
+  void recordScope(PartialApplyInst *PAI) {
+    // Only track scopes of non-escaping closures.
+    auto closureTy = PAI->getCallee()->getType().castTo<SILFunctionType>();
+    if (!isNonEscapingClosure(closureTy))
+      return;
+
+    auto closureFunc = PAI->getCalleeFunction();
+    assert(closureFunc && "non-escaping closure needs a direct partial_apply.");
+
+    auto scopeFunc = PAI->getFunction();
+    int scopeIdx = lookupScopeIndex(scopeFunc);
+
+    auto &indices = closureToScopesMap[closureFunc];
+    if (std::find(indices.begin(), indices.end(), scopeIdx) != indices.end())
+      return;
+
+    indices.push_back(scopeIdx);
+  }
+
+protected:
+  int lookupScopeIndex(SILFunction *scopeFunc) {
+    auto indexPos = scopeToIndexMap.find(scopeFunc);
+    if (indexPos != scopeToIndexMap.end())
+      return indexPos->second;
+
+    int scopeIdx = indexedScopes.size();
+    scopeToIndexMap[scopeFunc] = scopeIdx;
+    indexedScopes.push_back(scopeFunc);
+    return scopeIdx;
+  }
+};
+
+void ClosureScopeData::compute(SILModule *M) {
+  for (auto &F : *M) {
+    for (auto &BB : F) {
+      for (auto &I : BB) {
+        if (auto *PAI = dyn_cast<PartialApplyInst>(&I)) {
+          recordScope(PAI);
```

Comment:
Not a real comment, but an aside: Not saying that you should do this, but this code is one reason why I would love an inst_iterator or something like that. Imagine if this code could be:

```
for (auto *PAI : inst_iterator<PartialApplyInst>(F)) {
  recordScope(PAI);
}
```